test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java

```diff
@@ -0,0 +1,422 @@
+                + ", inv2=" + inv2);
+    }
+
+    public void checkEq(int returnValue) {
+        int invDiff = inv2 - inv1;
+        if ((invDiff < size && returnValue != invDiff) || (invDiff
+                    >= size && returnValue != size)) {
+            fail(returnValue);
+        }
+    }
+
+    public void checkNe(int returnValue) {
+        int invDiff = inv2 - inv1;
+        if ((invDiff != 0 && returnValue != 0) || (invDiff
+                    == 0 && returnValue != 1)) {
+            fail(returnValue);
+        }
+    }
+
+    @Test
+    @Arguments(setup = "setupEq")
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public int equalsAddInt(int inv1, int inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv2 - inv1 == i`
+            if (inv1 + i == inv2) {
+                blackhole();
+                break;
+            }
+        }
+        return i;
+    }
+
+    @Check(test = "equalsAddInt")
+    public void checkEqualsAddInt(int returnValue) {
+        checkEq(returnValue);
+    }
+
+    @Test
+    @Arguments(setup = "setupEq")
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public int equalsAddLong(long inv1, long inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv2 - inv1 == i`
+            if (inv1 + i == inv2) {
+                blackhole();
+                break;
+            }
+        }
+        return i;
+    }
+
+    @Check(test = "equalsAddLong")
+    public void checkEqualsAddLong(int returnValue) {
+        checkEq(returnValue);
+    }
+
+    @Test
+    @Arguments(setup = "setupEq")
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public int equalsInvariantSubVariantInt(int inv1, int inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv1 - inv2 == i`
+            if (inv2 - i == inv1) {
+                blackhole();
+                break;
+            }
+        }
+        return i;
+    }
+
+    @Check(test = "equalsInvariantSubVariantInt")
+    public void checkEqualsInvariantSubVariantInt(int returnValue) {
+        checkEq(returnValue);
+    }
+
+    @Test
+    @Arguments(setup = "setupEq")
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public int equalsInvariantSubVariantLong(long inv1, long inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv1 - inv2 == i`
+            if (inv2 - i == inv1) {
+                blackhole();
+                break;
+            }
+        }
+        return i;
+    }
+
+    @Check(test = "equalsInvariantSubVariantLong")
+    public void checkEqualsInvariantSubVariantLong(int returnValue) {
+        checkEq(returnValue);
+    }
+
+    @Test
+    @Arguments(setup = "setupEq")
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public int equalsVariantSubInvariantInt(int inv1, int inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv1 + inv2 == i`
```

Comment:
```suggestion
            // Reassociate to `inv2 - inv1 == i`
```
Or otherwise you'd have to explain why you do the `@IR(counts = {IRNode.SUB_I, "1"})` here